File: frontend/lib/taxonomy.ts
"use client";

import { useEffect, useState } from "react";
import { apiClient } from "@/lib/api/client";

/**
 * The five Category chips the design system's Add-item and Closet screens
 * both specify. `full_body` has no chip of its own and lives under Bottoms —
 * the same mapping feature 004 resolved for the Closet filter, kept
 * identical here so a dress is filed the same way in both places.
 */
export const CATEGORY_CHIPS = [
  { group: "top", label: "Top" },
  { group: "bottom", label: "Bottom" },
  { group: "outerwear", label: "Outerwear" },
  { group: "footwear", label: "Footwear" },
  { group: "accessory", label: "Accessory" },
] as const;

export type CategoryChip = (typeof CATEGORY_CHIPS)[number]["group"];

/** Bottoms covers `full_body`, so its type list is the union of both. */
export function groupsForChip(chip: CategoryChip): string[] {
  return chip === "bottom" ? ["bottom", "full_body"] : [chip];
}

/** `necklace` -> `Necklace`, `bow_tie` -> `Bow tie`, `t-shirt` -> `T-shirt`. */
export function humanizeCategory(value: string): string {
  const spaced = value.replace(/_/g, " ");
  return spaced.charAt(0).toUpperCase() + spaced.slice(1);
}

/**
 * `{group: [specific categories]}` from `GET /api/v1/taxonomy/categories`.
 *
 * Fetched rather than hard-coded: `categories.CATEGORY_GROUPS` is the single
 * source of truth and grows whenever new garment types show up in data. A
 * hand-mirrored copy of it would drift — this project already had one such
 * mirror (the colour palette, carrying a "keep in sync by hand" comment)
 * and deleted it once the review card stopped needing names at all
 * (constitution VII).
 *
 * Returns `{}` until it loads, and on failure — the Group control degrades
 * to free text, which is exactly what `category` is on the backend anyway,
 * so a taxonomy fetch failing never blocks adding an item.
 */
export function useCategoryTaxonomy(): Record<string, string[]> {
  const [taxonomy, setTaxonomy] = useState<Record<string, string[]>>({});

  useEffect(() => {
    let cancelled = false;
    (async () => {
      try {
        const { data } = await apiClient.GET("/api/v1/taxonomy/categories", {});
        if (!cancelled && data) setTaxonomy(data as Record<string, string[]>);
      } catch {
        // Left empty — see the docstring: the Group field still works.
      }
    })();
    return () => {
      cancelled = true;
    };
  }, []);

  return taxonomy;
}

File: backend/src/whattowear/api/v1/routes/taxonomy.py
"""GET /api/v1/taxonomy/categories — the category vocabulary, grouped.

Exists so the Add-item review card can offer the concrete garment types that
belong to a chosen category ("Accessory" -> tie, bow tie, necklace, ring)
without the frontend hand-mirroring `categories.CATEGORY_GROUPS`. That table
changes whenever new categories show up in data, and a hand-copy would drift
— the colour palette is already mirrored by hand with a "keep in sync"
comment, and one of those is enough (constitution VII).

Unauthenticated on purpose: it is a static, public vocabulary with nothing
user-specific in it, and the Add screen needs it before any per-user data
loads.
"""

from __future__ import annotations

from fastapi import APIRouter

from whattowear.categories import specifics_by_group

router = APIRouter()


@router.get("/taxonomy/categories")
def get_category_taxonomy() -> dict[str, list[str]]:
    """`{group: [specific categories]}`. The six group names themselves are
    excluded — they are the keys, and offering "top" as a type within "top"
    would be noise. A caller that genuinely has no better answer stores the
    bare group name, which `categories.group_of` round-trips."""
    return specifics_by_group()
